site/tests/app/controllers/course/CourseMaterialsControllerTester.php

```diff
@@ -0,0 +1,268 @@
+                    $zip->addFile( $this->config['course_path'] . $lev . "/" . $fname  );
+                    $files[] = $this->config['course_path'] . $lev . "/" . $fname;
+                }
+                $lev .= "/lev" . $i . "";
+                FileUtils::createDir($this->config['course_path'] . $lev);
+                
+            }
+
+            $zip->close();
+        }
+
+        $_FILES["files{$part}"]['name'][] = $name;
+        $_FILES["files{$part}"]['type'][] = FileUtils::getMimeType($filename_full);
+        $_FILES["files{$part}"]['size'][] = filesize($this->config['course_path'] . "/" .   $name);
+        
+        $tmpname = $this->config['course_path'] . "/" . Utils::generateRandomString() . $name;
+        copy($this->config['course_path'] . "/" . $name, $tmpname);
+
+        $_FILES["files{$part}"]['tmp_name'][] = $tmpname;
+        $_FILES["files{$part}"]['error'][] = 0;
+
+        return $files;
+    }
+
+    public function testCourseMaterialsUpload(){
+        $controller = new CourseMaterialsController($this->core);
+
+        $name = "foo.txt";
+        file_put_contents($this->upload_path . "/" .  $name, 'a');
+        $this->buildFakeFile($name);
+
+        $ret = $controller->ajaxUploadCourseMaterialsFiles();
+
+        $json = FileUtils::readJsonFile($this->json_path);
+        //we need to check that the file exists in the correct folder and also the JSON file
+        $filename_full = FileUtils::joinPaths( $this->upload_path, $name );
+        $expected_json = [ 
+            $filename_full => [
+                "checked" => 1,
+                "release_datetime" => $_POST['release_time']
+            ]
+        ];
+        $this->assertEquals($expected_json, $json);
+        //check the uploads directory now 
+        $files = FileUtils::getAllFiles($this->upload_path, array(), true);
+
+        $expected_files = [ 
+            $name => [
+                'name' => $name,
+                'path' => $filename_full,
+                'size' => filesize($filename_full),
+                'relative_name' => $name
+            ]
+        ];
+
+        $this->assertEquals($expected_files, $files);
+    }
+
+    public function testZipCourseUpload(){
+        $controller = new CourseMaterialsController($this->core);
+
+        $_FILES = array();
+        $_POST['expand_zip'] = 'on';
+        //create a zip file of depth = 2 with 2 files in each level.
+        $fake_files = $this->buildFakeZipFile('foo.zip', 1, 2, 2);
+        $ret = $controller->ajaxUploadCourseMaterialsFiles();
+        $json = FileUtils::readJsonFile($this->json_path);
+        
+        $expected_ret = array('status' => 'success', 'data' => 'Successfully uploaded!');
+        $this->assertEquals($expected_ret, $ret);
+
+        $files = FileUtils::getAllFiles($this->upload_path, array(), true);
+        $this->assertEquals(4, count($files));
+
+        $f1 = array_key_first($files);
```

Comment:
while a useful function, this is only available on PHP >= 7.3 (see header in https://www.php.net/manual/en/function.array-key-first.php). Please add this to `Utils` and use that:
```
public static function getFirstArrayElement($array) {
        foreach($array as $key => $unused) {
            return $key;
        }
        return NULL;
}
```

to match the existing `Utils::getLastArrayElement($array)`

Reply:
Thanks !